src/NHibernate/Multi/QueryBatchItemBase.cs

```diff
@@ -243,8 +294,15 @@ private void InitializeEntitiesAndCollections(DbDataReader reader, List<object>[
 				if (queryInfo.IsResultFromCache)
 					continue;
 				queryInfo.Loader.InitializeEntitiesAndCollections(
-					hydratedObjects[i], reader, Session, Session.PersistenceContext.DefaultReadOnly);
+					hydratedObjects[i], reader, Session, queryInfo.Parameters.IsReadOnly(Session),
```

Comment:
This change would also have been better moved in #1789. The old code was from multi-query/criteria, which was not knowing for which queries they were initializing and so could not handle their readonly flag.